(maint) Do not use Ruby stacktrace for "Puppet" stack

justinstoller opened the pull request:
Previously, we used `Kernel.eval` to add additional information to the
Ruby call stack. That information would show that the call had come from
a Puppet manifest. We did this to interleave Puppet "function calls"
with Ruby function calls. Then when creating error or warning messages
we would retrieve the entire Ruby call stack and strip out everything
from the call stack save for the Puppet manifest files.

This moves us to use a threadsafe variable to store Puppet stack
information bypassing the performance issues that come from interacting
with the Ruby call stack.

justinstoller commented on lib/puppet/pops/puppet_stack.rb:
```diff
@@ -1,57 +1,62 @@
+    # as well as a nil line number are supported. Such stack frames
+    # will be represented with the text `unknown` and `0´ respectively.
+    class PuppetStack
+      @@stack = Puppet::ThreadLocal.new { Array.new }
```
I don't like the class variable here, but wasn't sure if we should tie this to the context

pcarlisle replied:
Why not just a regular class instance variable though

justinstoller replied:
If I had a good reason. I don't remember it. What do you think about class vs module (see https://github.com/puppetlabs/puppet/pull/7754#discussion_r330304454)? If the answer is module that might remove the class vs class instance question?

justinstoller replied:
Moved this to a "class instance variable" or "module instance variable" or whatever it's called.

justinstoller commented on lib/puppet/pops/puppet_stack.rb:
```diff
@@ -1,57 +1,62 @@
+    # To support testing, a given file that is an empty string, or nil
+    # as well as a nil line number are supported. Such stack frames
+    # will be represented with the text `unknown` and `0´ respectively.
+    class PuppetStack
```
This was a module before...

I checked in our code base and stdlib and only found usages of it that called the module level methods directly on the module (which is also how we document its api), rather than including the module into another class.

This would break users who were including it, I don't know if we have any idea if any users are including it into other classes though....

justinstoller replied:
Moved this to be a module and it seems fine.

justinstoller commented on lib/puppet/pops/evaluator/runtime3_support.rb:
```diff
@@ -298,14 +298,11 @@ def external_call_function(name, args, scope, &block)
+    if func
       Puppet::Util::Profiler.profile(name, [:functions, name]) do
-        # Add stack frame when calling. See Puppet::Pops::PuppetStack
-        return Kernel.eval('_func.call(scope, *args, &block)'.freeze, Kernel.binding, file || '', line)
```
This was moved to not call PuppetStack.stack in PUP-6622 as a performance improvement on Windows. While I think this work in general should provide a meaningful performance speed up in JRuby, we should validate and that there wasn't a regression in Windows.

justinstoller commented:
Also need to run this (and benchmark) w/ https://github.com/puppetlabs/puppetlabs-stdlib/blob/master/lib/puppet/functions/deprecation.rb#L19 . I think @pcarlisle noticed that function hammering server in some code bases.

justinstoller commented:
A quick comparison between master:
```
16:37:08 puppet justin (master $<>) ‖ ITERATIONS=100 be rake benchmark:full_catalog 2>/dev/null
Fetching gem metadata from https://rubygems.org/..........
...snip...
Use `bundle info [gemname]` to see where a bundled gem is installed.
                 user     system      total        real
Run 1       14.110000   0.320000  14.430000 (  4.419816)
Run 2        9.780000   0.250000  10.030000 (  3.605717)
Run 3        6.980000   0.170000   7.150000 (  3.148417)
Run 4        7.620000   0.200000   7.820000 (  2.924873)
Run 5        5.900000   0.140000   6.040000 (  2.722796)
Run 6        6.050000   0.170000   6.220000 (  2.619804)
Run 7        4.800000   0.100000   4.900000 (  2.536014)
Run 8        4.040000   0.080000   4.120000 (  2.424491)
Run 9        5.070000   0.150000   5.220000 (  2.437842)
Run 10       4.320000   0.090000   4.410000 (  2.385887)
...snip...
Run 90       2.620000   0.060000   2.680000 (  1.966440)
Run 91       2.630000   0.060000   2.690000 (  1.970405)
Run 92       2.250000   0.050000   2.300000 (  1.892272)
Run 93       2.750000   0.060000   2.810000 (  1.944515)
Run 94       1.940000   0.040000   1.980000 (  1.866056)
Run 95       2.720000   0.060000   2.780000 (  1.971817)
Run 96       2.230000   0.050000   2.280000 (  1.883286)
Run 97       2.710000   0.060000   2.770000 (  1.980878)
Run 98       2.580000   0.050000   2.630000 (  1.982234)
Run 99       2.120000   0.050000   2.170000 (  1.880584)
Run 100      2.770000   0.050000   2.820000 (  1.962219)
> total:   358.700000   8.190000 366.890000 (211.974179)
> avg:       3.587000   0.081900   3.668900 (  2.119742)
```
And this PR + the caller improvements PR:
```
16:43:18 sellout:puppet justin (maint/master/puppetstack $) ‖ ITERATIONS=100 be rake benchmark:full_catalog 2>/dev/null
Fetching gem metadata from https://rubygems.org/..........
....snip....
Use `bundle info [gemname]` to see where a bundled gem is installed.
                 user     system      total        real
Run 1       15.750000   0.340000  16.090000 (  4.375356)
Run 2       10.200000   0.230000  10.430000 (  3.405501)
Run 3        8.750000   0.230000   8.980000 (  3.063337)
Run 4        6.880000   0.180000   7.060000 (  2.875986)
Run 5        6.320000   0.150000   6.470000 (  2.675978)
Run 6        5.560000   0.180000   5.740000 (  2.483797)
Run 7        5.710000   0.120000   5.830000 (  2.361031)
Run 8        4.310000   0.090000   4.400000 (  2.342954)
Run 9        4.510000   0.160000   4.670000 (  2.340585)
Run 10       4.440000   0.090000   4.530000 (  2.291779)
....snip....
Run 90       1.940000   0.040000   1.980000 (  1.783148)
Run 91       2.540000   0.060000   2.600000 (  1.854592)
Run 92       2.560000   0.060000   2.620000 (  1.864644)
Run 93       2.160000   0.050000   2.210000 (  1.796699)
Run 94       2.610000   0.050000   2.660000 (  1.857633)
Run 95       2.060000   0.050000   2.110000 (  1.771426)
Run 96       2.880000   0.060000   2.940000 (  1.861260)
Run 97       2.010000   0.050000   2.060000 (  1.781839)
Run 98       2.510000   0.060000   2.570000 (  1.857075)
Run 99       1.970000   0.040000   2.010000 (  1.784682)
Run 100      2.470000   0.060000   2.530000 (  1.846498)
> total:   360.730000   8.650000 369.380000 (204.705230)
> avg:       3.607300   0.086500   3.693800 (  2.047052)
```

It's not huge, but maybe not completely within the margin of error?

justinstoller commented:
The function loading benchmark mentioned in PUP-6622 seems problematic:
```
16:48:13 puppet justin (maint/master/puppetstack $) ‖ be rake benchmark:function_loading
mkdir -p /var/folders/8t/njts_3s96_v6f30y1b8b92c00000gn/T/function_loading20191001-49792-g6jlm0
                 user     system      total        real
Run 1      /Users/justin/Backup/code/work/puppet/.bundle/gems/gems/rake-12.3.3/lib/rake/task.rb:225: warning: singleton on non-persistent Java type Java::JavaLang::StackOv
erflowError (http://wiki.jruby.org/Persistence)
rake aborted!
Java::JavaLang::StackOverflowError:
Users.justin.Backup.code.work.puppet.lib.puppet.pops.loader.base_loader.invokeOther6:load_typed(/Users/justin/Backup/code/work/puppet/lib/puppet/pops/loader/base_loader.rb
:155)
Users.justin.Backup.code.work.puppet.lib.puppet.pops.loader.base_loader.RUBY$method$internal_load$0(/Users/justin/Backup/code/work/puppet/lib/puppet/pops/loader/base_loade
r.rb:155)
org.jruby.internal.runtime.methods.CompiledIRMethod.call(CompiledIRMethod.java:117)
org.jruby.internal.runtime.methods.MixedModeIRMethod.call(MixedModeIRMethod.java:156)
org.jruby.internal.runtime.methods.DynamicMethod.call(DynamicMethod.java:200)
org.jruby.runtime.callsite.CachingCallSite.cacheAndCall(CachingCallSite.java:375)
org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:174)
Users.justin.Backup.code.work.puppet.lib.puppet.pops.loader.base_loader.invokeOther5:internal_load(/Users/justin/Backup/code/work/puppet/lib/puppet/pops/loader/base_loader
.rb:42)
Users.justin.Backup.code.work.puppet.lib.puppet.pops.loader.base_loader.RUBY$method$load_typed$0(/Users/justin/Backup/code/work/puppet/lib/puppet/pops/loader/base_loader.r
b:42)
```

justinstoller commented:
I still have to check on a Windows machine but locally these changes seem to do no harm:

master:
```
19:47:16 puppet justin (master $<>) ‖ ITERATIONS=20 be rake benchmark:function_loading
mkdir -p /var/folders/8t/njts_3s96_v6f30y1b8b92c00000gn/T/function_loading20191001-53391-1e6h82l
                 user     system      total        real
Run 1        5.009273   0.255079   5.280144 (  5.300357)
Run 2        4.996746   0.206641   5.203387 (  5.207923)
Run 3        4.941208   0.220288   5.161496 (  5.177845)
Run 4        4.924104   0.203146   5.127250 (  5.136081)
Run 5        4.924354   0.224665   5.149019 (  5.155071)
Run 6        5.132853   0.214228   5.347081 (  5.365698)
Run 7        5.754574   0.211301   5.965875 (  6.014642)
Run 8        5.258642   0.198008   5.456650 (  5.458846)
Run 9        4.448036   0.220268   4.668304 (  4.668604)
Run 10       5.531215   0.188847   5.720062 (  5.723464)
Run 11       5.711594   0.205594   5.917188 (  5.926093)
Run 12       4.582950   0.232555   4.815505 (  4.818813)
Run 13       5.511823   0.199205   5.711028 (  5.718751)
Run 14       5.186403   0.193605   5.380008 (  5.384509)
Run 15       5.617085   0.196417   5.813502 (  5.815900)
Run 16       5.603910   0.249082   5.852992 (  5.864056)
Run 17       5.859642   0.220360   6.080002 (  6.089170)
Run 18       4.369657   0.192504   4.562161 (  4.564382)
Run 19       5.757970   0.200814   5.958784 (  5.961773)
Run 20       4.542935   0.188610   4.731545 (  4.737982)
> total:   103.664974   4.221217 107.901983 (108.089960)
> avg:       5.183249   0.211061   5.395099 (  5.404498)
```

With the this pr + the caller improvements, as I tested with JRuby:

```
19:49:24 puppet justin (master $<>) ‖ gk -
Switched to branch 'maint/master/puppetstack'

19:50:22 puppet justin (maint/master/puppetstack $) ‖ ITERATIONS=20 be rake benchmark:function_loading
mkdir -p /var/folders/8t/njts_3s96_v6f30y1b8b92c00000gn/T/function_loading20191001-53532-5a9lxi
                 user     system      total        real
Run 1        3.759157   0.218451   3.993743 (  4.000105)
Run 2        3.829885   0.196607   4.026492 (  4.029216)
Run 3        3.678133   0.197608   3.875741 (  3.876467)
Run 4        3.861066   0.190409   4.051475 (  4.053027)
Run 5        4.021655   0.202617   4.224272 (  4.231858)
Run 6        4.276638   0.223379   4.500017 (  4.510991)
Run 7        4.322435   0.221271   4.543706 (  4.557815)
Run 8        4.487837   0.227327   4.715164 (  4.731206)
Run 9        4.056145   0.218808   4.274953 (  4.277837)
Run 10       3.666868   0.191269   3.858137 (  3.863616)
Run 11       4.190758   0.193647   4.384405 (  4.386642)
Run 12       4.382706   0.207346   4.590052 (  4.594156)
Run 13       4.353183   0.209831   4.563014 (  4.567540)
Run 14       4.183699   0.183081   4.366780 (  4.371759)
Run 15       5.209367   0.218052   5.427419 (  5.443028)
Run 16       3.540611   0.233349   3.773960 (  3.797889)
Run 17       4.957908   0.221932   5.179840 (  5.199488)
Run 18       4.246186   0.205165   4.451351 (  4.468557)
Run 19       4.275672   0.200842   4.476514 (  4.490196)
Run 20       4.700806   0.199747   4.900553 (  4.902566)
> total:    84.000715   4.160738  88.177588 ( 88.353959)
> avg:       4.200036   0.208037   4.408879 (  4.417698)
```

puppetcla commented:
CLA signed by all contributors.

justinstoller commented:
It looks like perf-wise this is working well on both MRI and JRuby. I removed the hyperbole from the commit message.

I'm still unsure if I should keep the class/class variable approach or move to using a module w/ context.

Magisus commented:
Do you know why this was originally using `Kernal.eval`? The original commit makes no indication of why that choice was made.

Magisus commented on lib/puppet/pops/puppet_stack.rb:
```diff
@@ -1,57 +1,62 @@
-    end
-  end
+      def self.stack(file, line, obj, message, args, &block)
+        #binding.pry
```
This got left behind

ciprianbadescu commented:
jenkins please test this on
debian8-64a,debian8-32a,debian9-64a,debian9-32a,fedora28-64a,fedora29-64a,osx1012-64a,osx1013-64a,osx1014-64a,redhat5-64a,centos6-64a,centos6-32a,redhat7-64a,redhat8-64a,redhatfips7-64a,sles11-64a,sles11-32a,sles12-64a,sles12-POWERa,sles15-64a,solaris10-64a,solaris114-64a,ubuntu140464a,ubuntu1404-32a,ubuntu1604-64a,ubuntu1604-32a,ubuntu1804-64a,windows2016-64a,windows2019-64a

justinstoller commented:
> Do you know why this was originally using Kernal.eval? 

For some reason I had it in my head that when working on the 4x compiler it was nice to see a Ruby stacktrace with Puppet code "call sites" w/in it and then that implementation provided an easy way to create Puppet "stacktraces". That's complete hearsay at this point though (I don't remember why I think that and I wasn't actually involved in the original implementation).

justinstoller commented:
Are there any blockers to this besides the comment?

pcarlisle commented:
This essentially fixes the performance problems with heavy use of the stdlib deprecation function (in my benchmark avg 4.4 second compilation times without this PR, 2.6 seconds with this PR)

justinstoller commented:
Moved back to a module and updated to use a "class" instance variable. Any blockers to merging this?
/cc @pcarlisle , @joshcooper

justinstoller commented:
Is it still possible to get this in for 6.11?

justinstoller commented:
jenkins please test this on redhat7-64a,windows2019-64a with servertests

justinstoller commented:
Here's a trivial usage of printing a Puppet Stack w/ v4 functions:
```
11:31:50 puppet justin (maint/master/puppetstack $%) ‖ be puppet apply test.pp 
Warning: Facter: Could not find a default route. Using first non-loopback interface
Error: Evaluation Error: Error while evaluating a Function Call,  (file: /Users/justin/Backup/code/work/puppet/test.pp, line: 1, column: 1) on node localhost

11:31:56 puppet justin (maint/master/puppetstack $%) ‖ gk master
Switched to branch 'master'
Your branch is up to date with 'puppet/master'.

11:32:09 puppet justin (master $%<>) ‖ be puppet apply test.pp 
Warning: Facter: Could not find a default route. Using first non-loopback interface
Error: Evaluation Error: Error while evaluating a Function Call,  (file: /Users/justin/Backup/code/work/puppet/test.pp, line: 1, column: 1) on node localhost
```

Where the bad behavior is defined in this .pp file:
```
11:32:15 puppet justin (master $%<>) ‖ cat test.pp 
broken("What")

11:34:12 puppet justin (master $%<>) ‖ cat lib/puppet/functions/broken.rb 
# Logs a message on the server at level `crit`.
Puppet::Functions.create_function(:broken, Puppet::Functions::InternalFunction) do
  # @param values The values to log.
  # @return [Undef]
  dispatch :broken do
    scope_param
    repeated_param 'Any', :values
    return_type 'Undef'
  end

  def broken(scope, *values)
    raise
  end
end
```

And the same behavior with v3 functions:
```
11:39:48 puppet justin (master $%<>) ‖ be puppet apply test.pp 
Warning: Facter: Could not find a default route. Using first non-loopback interface
Error: Evaluation Error: Error while evaluating a Function Call,  (file: /Users/justin/Backup/code/work/puppet/test.pp, line: 1, column: 1) on node localhost

11:39:54 puppet justin (master $%<>) ‖ gk -
Switched to branch 'maint/master/puppetstack'

11:41:09 puppet justin (maint/master/puppetstack $%) ‖ be puppet apply test.pp 
Warning: Facter: Could not find a default route. Using first non-loopback interface
Error: Evaluation Error: Error while evaluating a Function Call,  (file: /Users/justin/Backup/code/work/puppet/test.pp, line: 1, column: 1) on node localhost

11:41:15 puppet justin (maint/master/puppetstack $%) ‖ cat test.pp 
broken_foo("What")

11:42:37 puppet justin (maint/master/puppetstack $%) ‖ cat lib/puppet/parser/functions/broken_foo.rb 
Puppet::Parser::Functions::newfunction(
    :broken_foo,
    :arity => -1,
    :doc   => <<DOC
Fail with a parse error. Any parameters will be stringified,
concatenated, and passed to the exception-handler.
DOC
) do |vals|
    raise
    vals = vals.collect { |s| s.to_s }.join(" ") if vals.is_a? Array
    raise Puppet::ParseError, vals.to_s
end
```

justinstoller commented:
@joshcooper also reminded me to get updated windows benchmarks regarding the change in runtime3_support.rb/PUP-6622/benchmark:function_loading. Here's some screenshots. It looks like there's no performance degradation from removing the direct Kernel.eval in runtime3_support and potentially a perf improvement overall for functions.

<img width="1440" alt="Screen Shot 2019-11-13 at 3 30 01 PM" src="https://user-images.githubusercontent.com/241423/68813684-6d940900-062b-11ea-8cd3-851b78ce1982.png">

<img width="1440" alt="Screen Shot 2019-11-13 at 3 33 39 PM" src="https://user-images.githubusercontent.com/241423/68813699-7389ea00-062b-11ea-8255-ae7b4b833f19.png">

hlindberg commented:
How come this change is made as a `(maint)` - there should have been a ticket!

justinstoller commented:
Very fair.

If you're actually curious, this was something that I found and toyed with in some free time and forgot about. I found it when cleaning out some old branches and put it up. But regardless, I should have made a ticket for this kind of change.
